File: desktop/src/api/task.ts
import { request, unwrap } from "./request";

export interface ServerTaskDto {
  id: number;
  user_id: number;
  title: string;
  content: string | null;
  status: string;
  priority: number;
  tag: string | null;
  project?: string | null;
  list_type?: string | null;
  due_time: string | null;
  remind_time: string | null;
  repeat_rule: string | null;
  planned_date?: string | null;
  completed_at?: string | null;
  snoozed_until?: string | null;
  parent_task_id?: number | null;
  checklist?: Array<{ id: string; title: string; done: boolean }>;
  is_template?: boolean;
  template_name?: string | null;
  sort_order?: number;
  version: number;
  is_deleted: boolean;
  created_at: string;
  updated_at: string;
  deleted_at: string | null;
}

export interface TaskCreatePayload {
  title: string;
  content?: string | null;
  priority?: number;
  tag?: string | null;
  project?: string | null;
  list_type?: string | null;
  due_time?: string | null;
  remind_time?: string | null;
  repeat_rule?: string | null;
  planned_date?: string | null;
  completed_at?: string | null;
  snoozed_until?: string | null;
  parent_task_id?: number | null;
  checklist?: Array<{ id: string; title: string; done: boolean }>;
  is_template?: boolean;
  template_name?: string | null;
  sort_order?: number;
}

export interface TaskUpdatePayload extends Partial<TaskCreatePayload> {
  status?: string;
  expected_version?: number;
}

export interface TaskListQuery {
  q?: string;
  view?: string;
  now?: string;
  status?: string;
  tag?: string;
  project?: string;
  listType?: string;
  plannedDate?: string;
  includeDeleted?: boolean;
  templatesOnly?: boolean;
  cursorId?: number;
  cursorUpdatedAt?: string;
  offset?: number;
  limit?: number;
  timezone?: string;
}

export interface TaskMetaDto {
  projects: string[];
  tags: string[];
  counts: {
    open: number;
    completed: number;
    inbox: number;
    today: number;
    overdue: number;
    templates: number;
    trash: number;
  };
}

export interface ChecklistItemPayload {
  id: string;
  title: string;
  done: boolean;
}

export interface ChecklistItemUpdatePayload {
  title?: string | null;
  done?: boolean | null;
}

export interface TaskHistoryDto {
  id: number;
  task_id: number | null;
  operation: string;
  result: string;
  version: number;
  device_id: string | null;
  local_id: string | null;
  server_id: number | null;
  payload: Record<string, unknown> | null;
  created_at: string;
}

export interface TemplateInstantiatePayload {
  title?: string | null;
  content?: string | null;
  project?: string | null;
  tag?: string | null;
  list_type?: string | null;
  due_time?: string | null;
  remind_time?: string | null;
  planned_date?: string | null;
}

export function fetchTasks(query: TaskListQuery = {}): Promise<ServerTaskDto[]> {
  const params = taskListQueryParams(query);
  return unwrap(request.get("/tasks", { params }));
}

export function fetchTask(taskId: number): Promise<ServerTaskDto> {
  return unwrap(request.get(`/tasks/${taskId}`));
}

export function fetchTaskMeta(timezone: string, now?: string): Promise<TaskMetaDto> {
  const params = new URLSearchParams();
  params.set("timezone", timezone.trim());
  appendStringParam(params, "now", now);
  return unwrap(request.get("/tasks/meta", { params: toRequestParams(params) }));
}

export function fetchTaskHistory(taskId: number): Promise<TaskHistoryDto[]> {
  return unwrap(request.get(`/tasks/${taskId}/history`));
}

export function createTask(payload: TaskCreatePayload): Promise<ServerTaskDto> {
  return unwrap(request.post("/tasks", payload));
}

export function updateTask(taskId: number, payload: TaskUpdatePayload): Promise<ServerTaskDto> {
  return unwrap(request.put(`/tasks/${taskId}`, payload));
}

export function deleteTask(taskId: number, expectedVersion?: number): Promise<ServerTaskDto> {
  return unwrap(request.delete(`/tasks/${taskId}`, { params: expectedVersionParams(expectedVersion) }));
}

export function purgeTask(taskId: number): Promise<ServerTaskDto> {
  return unwrap(request.delete(`/tasks/${taskId}/purge`));
}

export function addChecklistItem(taskId: number, payload: ChecklistItemPayload): Promise<ServerTaskDto> {
  return unwrap(request.post(`/tasks/${taskId}/checklist`, payload));
}

export function updateChecklistItem(
  taskId: number,
  itemId: string,
  payload: ChecklistItemUpdatePayload,
): Promise<ServerTaskDto> {
  return unwrap(request.put(`/tasks/${taskId}/checklist/${itemId}`, payload));
}

export function deleteChecklistItem(taskId: number, itemId: string): Promise<ServerTaskDto> {
  return unwrap(request.delete(`/tasks/${taskId}/checklist/${itemId}`));
}

export function completeTask(taskId: number, expectedVersion?: number): Promise<ServerTaskDto> {
  return unwrap(request.post(`/tasks/${taskId}/complete`, undefined, { params: expectedVersionParams(expectedVersion) }));
}

export function restoreTask(taskId: number, expectedVersion?: number): Promise<ServerTaskDto> {
  return unwrap(request.post(`/tasks/${taskId}/restore`, undefined, { params: expectedVersionParams(expectedVersion) }));
}

export function createNextOccurrence(taskId: number): Promise<ServerTaskDto> {
  return unwrap(request.post(`/tasks/${taskId}/next-occurrence`));
}

export function instantiateTemplate(
  templateId: number,
  payload: TemplateInstantiatePayload,
): Promise<ServerTaskDto> {
  return unwrap(request.post(`/tasks/templates/${templateId}/instantiate`, payload));
}

function expectedVersionParams(expectedVersion?: number): Record<string, unknown> {
  return Number.isFinite(expectedVersion) ? { expected_version: expectedVersion } : {};
}

function taskListQueryParams(query: TaskListQuery): Record<string, unknown> {
  const params = new URLSearchParams();
  appendStringParam(params, "q", query.q);
  appendStringParam(params, "view", query.view);
  appendStringParam(params, "now", query.now);
  appendStringParam(params, "status", query.status);
  appendStringParam(params, "tag", query.tag);
  appendStringParam(params, "project", query.project);
  appendStringParam(params, "list_type", query.listType);
  appendStringParam(params, "planned_date", query.plannedDate);
  appendBooleanParam(params, "include_deleted", query.includeDeleted);
  appendBooleanParam(params, "templates_only", query.templatesOnly);
  appendNumberParam(params, "cursor_id", query.cursorId);
  appendStringParam(params, "cursor_updated_at", query.cursorUpdatedAt);
  appendNumberParam(params, "offset", query.offset);
  appendNumberParam(params, "limit", query.limit);
  if (query.timezone?.trim()) params.set("timezone", query.timezone.trim());
  return toRequestParams(params);
}

function toRequestParams(params: URLSearchParams): Record<string, unknown> {
  return Object.fromEntries(params.entries());
}

function appendStringParam(params: URLSearchParams, key: string, value: string | undefined): void {
  const trimmed = value?.trim();
  if (trimmed) params.set(key, trimmed);
}

function appendBooleanParam(params: URLSearchParams, key: string, value: boolean | undefined): void {
  if (value !== undefined) params.set(key, String(value));
}

function appendNumberParam(params: URLSearchParams, key: string, value: number | undefined): void {
  if (Number.isFinite(value)) params.set(key, String(value));
}

File: backend/app/models/task.py
from datetime import date, datetime

from sqlalchemy import Boolean, Date, DateTime, ForeignKey, Index, Integer, JSON, String, Text
from sqlalchemy.orm import Mapped, mapped_column

from app.core.database import Base
from app.utils.time import utc_now


class Task(Base):
    __tablename__ = "tasks"
    __table_args__ = (
        Index("ix_tasks_user_deleted_updated", "user_id", "is_deleted", "updated_at"),
        Index("ix_tasks_user_status", "user_id", "status"),
        Index("ix_tasks_user_tag", "user_id", "tag"),
        Index("ix_tasks_user_template", "user_id", "is_template"),
        Index("ix_tasks_user_planned_status", "user_id", "planned_date", "status"),
    )

    id: Mapped[int] = mapped_column(primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id"), index=True, nullable=False)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    content: Mapped[str | None] = mapped_column(Text, nullable=True)
    status: Mapped[str] = mapped_column(String(32), default="todo", nullable=False)
    priority: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    tag: Mapped[str | None] = mapped_column(String(64), nullable=True)
    project: Mapped[str | None] = mapped_column(String(128), nullable=True)
    list_type: Mapped[str] = mapped_column(String(32), default="inbox", nullable=False)
    due_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=False), nullable=True)
    remind_time: Mapped[datetime | None] = mapped_column(DateTime(timezone=False), nullable=True)
    repeat_rule: Mapped[str | None] = mapped_column(String(255), nullable=True)
    planned_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    completed_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=False), nullable=True)
    snoozed_until: Mapped[datetime | None] = mapped_column(DateTime(timezone=False), nullable=True)
    parent_task_id: Mapped[int | None] = mapped_column(ForeignKey("tasks.id"), nullable=True)
    checklist: Mapped[list[dict] | None] = mapped_column(JSON, default=list, nullable=True)
    is_template: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    template_name: Mapped[str | None] = mapped_column(String(128), nullable=True)
    sort_order: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    version: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    is_deleted: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=False), default=utc_now)
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=False),
        default=utc_now,
        onupdate=utc_now,
    )
    deleted_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=False), nullable=True)
